File: docker-secondary-agent/src/structs/check_message.rs
use serde::*;

#[derive(Serialize, Deserialize)]
pub enum CheckBlockType {
    #[serde(rename = "code")]
    Code,
    #[serde(rename = "message")]
    Message,
}

#[derive(Serialize, Deserialize)]
pub struct CheckMessageBlock {
    #[serde(rename = "type")]
    pub block_type: CheckBlockType,
    pub content: String,
}

impl CheckMessageBlock {
    pub fn new_message(content: String) -> Self {
        CheckMessageBlock {
            content,
            block_type: CheckBlockType::Message,
        }
    }

    pub fn new_code(content: String) -> Self {
        CheckMessageBlock {
            content,
            block_type: CheckBlockType::Code,
        }
    }
}

#[derive(Serialize, Deserialize)]
pub struct CheckMessage {
    pub title: String,
    pub blocks: Vec<CheckMessageBlock>,
}

impl Default for CheckMessage {
    fn default() -> Self {
        CheckMessage {
            title: String::new(),
            blocks: Vec::new(),
        }
    }
}

impl CheckMessage {
    pub fn new(title: String) -> Self {
        CheckMessage {
            title,
            blocks: Vec::new(),
        }
    }
}

File: main-site/src/Data/Models/CheckMessage.hs
{-# LANGUAGE LambdaCase        #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}
module Data.Models.CheckMessage
  ( CheckMessage(..)
  , CheckMessageBlock(..)
  , CheckBlockType(..)
  ) where

import           Data.Aeson
import           Data.Text  (Text)

data CheckBlockType = Code | Message deriving Show

instance ToJSON CheckBlockType where
  toJSON Code    = String "code"
  toJSON Message = String "message"

instance FromJSON CheckBlockType where
  parseJSON = withText "CheckBlockType" $ \case
    "code" -> pure Code
    "message" -> pure Message
    _ -> fail "Invalid CheckBlockType!"

data CheckMessageBlock = CheckMessageBlock
  { getBlockType    :: !CheckBlockType
  , getBlockContent :: !Text
  } deriving Show

instance ToJSON CheckMessageBlock where
  toJSON (CheckMessageBlock { .. }) = object [ "type" .= getBlockType, "content" .= getBlockContent ]

instance FromJSON CheckMessageBlock where
  parseJSON = withObject "CheckMessageBlock" $ \v -> CheckMessageBlock
    <$> v .: "type"
    <*> v .: "content"

data CheckMessage = CheckMessage
  { getMessageTitle  :: !Text
  , getMessageBlocks :: ![CheckMessageBlock]
  } deriving Show

instance ToJSON CheckMessage where
  toJSON (CheckMessage { .. }) = object [ "title" .= getMessageTitle, "blocks" .= getMessageBlocks ]

instance FromJSON CheckMessage where
  parseJSON = withObject "CheckMessage" $ \v -> CheckMessage
    <$> v .: "title"
    <*> v .: "blocks"
